### PR TITLE
For #20077: reland async navgraph & address bugs

### PR DESCRIPTION
**Pull request by mcomella**
Running the UI tests against this PR to see what fails.

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by mcomella** [comment]
The UI tests passed on `run-ui (28, google_apis)` but failed one test, `TabbedBrowsingTest.verifyContextMenuShortcuts`, on `ui-test-x86-debug` with both failures occurring on the same spot:

```
androidx.test.espresso.PerformException: Error performing 'single click - At Coordinates: 359, 1710 and precision: 16, 16' on view 'with id: org.mozilla.fenix.debug:id/toolbar'.
at androidx.test.espresso.PerformException$Builder.build(PerformException.java:5)
at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:25)
at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:36)
at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:106)
at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:43)
at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:94)
at org.mozilla.fenix.helpers.ViewInteractionKt.click(ViewInteraction.kt:18)
at org.mozilla.fenix.ui.robots.NavigationToolbarRobotKt.openEditURLView(NavigationToolbarRobot.kt:312)
at org.mozilla.fenix.ui.robots.NavigationToolbarRobot$Transition.enterURLAndEnterToBrowser(NavigationToolbarRobot.kt:101)
at org.mozilla.fenix.ui.TabbedBrowsingTest.verifyContextMenuShortcuts(TabbedBrowsingTest.kt:328) 
```

Thoughts:
* Running on google may only run a subset of the tests which is why this didn't fail there
* This may have failed on the emulator because it's probably running in software mode and runs significantly slower there and could uncover race conditions. I'd be curious if this test also fails intermittently on master

**Comment by mcomella** [comment]
Let's rerun the tests to see if it fails consistently.

**Comment by mcomella** [comment]
I was able to reproduce the test failure locally and confirmed it was introduced by the changes in this PR.

**Review by MarcLeclair — CHANGES_REQUESTED** [review]
I think we should use the default dispatcher. Otherwise it looks good to me and it works too 👍

File: app/src/main/java/org/mozilla/fenix/perf/NavGraphProvider.kt [hunk 1]
```diff
@@ -30,15 +45,42 @@ object NavGraphProvider {
...
+    private data class DeferredNavGraphContainer(
+        val inflationJob: Deferred<NavGraph>?,
+        var isGraphSet: Boolean = false
+    )
+
+    @OptIn(ExperimentalCoroutinesApi::class) // UNDISPATCHED.
+    fun inflateNavGraphAsync(
+        navController: NavController,
+        lifecycleScope: CoroutineScope,
+        ioDispatcher: CoroutineDispatcher = Dispatchers.IO
```
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/perf/NavGraphProvider.kt:59]: @MarcLeclair AIUI, inflation reads the layout file from disk and then heavily uses CPU to complete the process. Going from this simple point of view, I think the IO dispatcher is more fitting.

From a more nuanced perspective, we should be careful about using the `Default` dispatcher for urgent operations because if the `Default` dispatcher gets too many jobs, newer jobs are blocked. I'd consider inflating the core UI very urgent so I'm hesitant to use it. I think if the Default dispatcher was implemented with more nuance, it might be an option. Specifically, if you're going to have a limited number of simultaneous jobs being executed (which is optimal for CPU-bound work), the jobs should be able to specify their relative priorities. For example, inflating the UI might be P1 and it'd bump any active P2 jobs from the queue until it completed. But that's not how the Default dispatcher is implemented so it's of limited utility.

What do you think?

[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/perf/NavGraphProvider.kt:59]: Spoke with MarcLeclair: we're okay with the IO dispatcher. MarcLeclair raised a point that we should try to set a good example by not defaulting to the IO dispatcher but we agreed unless we came up with a standard threading model for the app, it would be difficult to not make the IO dipatcher the default given the lack of a priority queue in the Default dispatcher.

tldr; we're fine with this change.

**Review by MarcLeclair — APPROVED** [review]
Looks good to me. Works and still have the perf wins on my G5+

**Comment by mcomella** [comment]
Rebasing – let's see how CI does.

**Comment by mcomella** [comment]
The unit test failure seems to be an intermittent related to https://github.com/mozilla-mobile/fenix/issues/20533#issuecomment-887851773 which seems to stem from the Java 11 upgrade:
```
io.mockk.MockKException: Can't instantiate proxy for class kotlin.Function2
	at io.mockk.impl.instantiation.JvmMockFactory.newProxy(JvmMockFactory.kt:64)
```

I'm going to rerun and see if it goes away.

---

UI test failure is in `org.mozilla.fenix.ui.SettingsPrivacyTest.launchLinksInPrivateToggleOffStateDoesntChangeTest` where both shards return the same error so it's possible it's a real error:
```
androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with content description text: is "Home screen" 
```

The test failure occurred locally but with a different error:
```
androidx.test.uiautomator.UiObjectNotFoundException: UiSelector[SCROLLABLE=true]
at androidx.test.uiautomator.UiScrollable.scrollBackward(UiScrollable.java:496)
at androidx.test.uiautomator.UiScrollable.scrollToBeginning(UiScrollable.java:544)
at androidx.test.uiautomator.UiScrollable.scrollToBeginning(UiScrollable.java:563)
at androidx.test.uiautomator.UiScrollable.scrollIntoView(UiScrollable.java:265)
at androidx.test.uiautomator.UiScrollable.scrollTextIntoView(UiScrollable.java:336)
at org.mozilla.fenix.helpers.TestHelper.scrollToElementByText(TestHelper.kt:52)
at org.mozilla.fenix.ui.robots.SettingsRobot$Transition.openPrivateBrowsingSubMenu(SettingsRobot.kt:209)
at org.mozilla.fenix.ui.SettingsPrivacyTest.launchLinksInPrivateToggleOffStateDoesntChangeTest(SettingsPrivacyTest.kt:356)
```

**Comment by mcomella** [comment]
The local test seems to fail because the settings menu is not opened right before `openPrivateBrowsingSubMenu` is called https://searchfox.org/mozilla-mobile/rev/becaeeb6fe4268631ff8155be9f3662917ee1bb3/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt#356 so `openPrivateBrowsingSubMenu` throws an assertion error because it doesn't see the text it's supposed to see. I think the reason the Settings menu isn't opened is because the 3-dot menu now requires a click and two swipes up to reach the bottom – perhaps the automated clicks are occurring too quickly.

**Comment by mcomella** [comment]
I filed https://github.com/mozilla-mobile/fenix/issues/20540 for the local test failure because it doesn't seem related to my change. After addressing the issue, I can't reproduce the UI test failure from CI so I'll rerun the tests to see if it's an intermittent.

**Comment by mcomella** [comment]
The same test failed with the same error (in addition to another flaky test failure). I guess I'll have to debug it.

**Comment by mcomella** [comment]
Watching [the video of the test failure](https://console.firebase.google.com/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/6590428936154009963/executions/bs.7b57a7de10eaf970/videos), we:
- Add the homescreen shortcut
- Open the app (presumably this is via the homescreen shortcut)

But instead of opening the page the homescreen shortcut points to, the homescreen. Since there is no homescreen button on the homescreen, this causes the "Home screen" button to be missing and an error is thrown.

I can't reproduce this issue locally: it's unclear if this is a real async nav graph issue (e.g. maybe the test device runs slowly, it causes unexpected behavior) or an intermittent failure. Given the fact that the main branch isn't failing with this, it seems like it could be a nav graph issue. Then again, all of the tests don't seem to be running in CI: https://github.com/mozilla-mobile/fenix/issues/20386

**Comment by mcomella** [comment]
This test passed in a recent PR I made against main https://github.com/mozilla-mobile/fenix/pull/20541/checks?check_run_id=3176544688 so it looks like it's the fault of this PR.

**Comment by mcomella** [comment]
I'm curious why this UI test passed before but doesn't pass now. I reverted some changes to that UI test from around Jul 12 so let's see if that fixes the issue. With that info, it might help nail down why the UI test is failing.

**Comment by mcomella** [comment]
The UI tests passed by reverting the subsequent changes: maybe that will help us figure out what happened.

**Comment by mcomella** [comment]
I was able to reproduce the UI test failure on a Pixel 2 API 28 virtual device (as opposed to my physical Pixel 2 API 30).

**Comment by mcomella** [comment]
I was not able to reproduce on the virtual device before my change so this change is in fact introduced by the nav graph. I'm guessing it has something to do with the fact that we're opening links in private browsing and that is expected to restart the activity.

**Comment by mcomella** [comment]
One discrepancy is that before the test failure:
- the test that passes calls `HomeActivity.onCreate` once
- the test that fails calls `HomeActivity.onCreate` twice

It's unclear why.

**Comment by mcomella** [comment]
It appears both sets of code [call `activity.recreate()` twice](https://searchfox.org/mozilla-mobile/rev/cb503a6fd0b5557a55d4117160da935f92b31de3/fenix/app/src/main/java/org/mozilla/fenix/theme/ThemeManager.kt#125) but the version of the test that passes (physical device API 30) only actually calls `HomeActivity.onCreate` once.

If I add a boolean to only allow one `recreate` call, both versions of the test fail.

---

On a side note, there are some wacky things in the code:
- I feel like the tab should be opened in PRIVATE mode because open links in PB is on but it's opened in normal mode (presumably because the tab is already opened in normal browsing)
- we [mutate `activity.intent` to set the private browsing state for the recreated activity](https://searchfox.org/mozilla-mobile/rev/cb503a6fd0b5557a55d4117160da935f92b31de3/fenix/app/src/main/java/org/mozilla/fenix/theme/ThemeManager.kt#122) but that intent doesn't represent the intent we just received: the `VIEW` from the homescreen shortcut

It feels like our PB mechanism is fragile. I'm concerned how the async nav graph interacts with these activity restarts, especially when they don't seem super consistent (e.g. calling recreate twice before it recreates).

**Comment by mcomella** [comment]
I tried to resummarize the issues I'm running into in the issue: https://github.com/mozilla-mobile/fenix/issues/20077#issuecomment-889507154

**Comment by mcomella** [comment]
Closing as inactive: see https://github.com/mozilla-mobile/fenix/issues/20077#issuecomment-897196279
